python/private/config_settings.bzl

```diff
@@ -0,0 +1,134 @@
+
+_PYTHON_VERSION = str(Label("//python/config_settings:python_version"))
+
+def _ver_key(s):
+    major, _, s = s.partition(".")
+    minor, _, s = s.partition(".")
+    micro, _, s = s.partition(".")
+    return (int(major), int(minor), int(micro))
+
+def _flag_values(python_versions):
+    """Construct a map of python_version to a list of toolchain values.
+
+    This mapping maps the concept of a config setting to a list of compatible toolchain versions.
+    For using this in the code, the VERSION_FLAG_VALUES should be used instead.
+
+    Args:
+        python_versions: A list of all versions.
+
+    Returns:
+        A map with config settings as keys and values as extra flag values to be included in
+        the config_setting_group if they should be also matched, which is used for generating
+        correct entries for matching the latest 3.8 version, etc.
+    """
+
+    # Maps e.g. "3.8" -> ["3.8.1", "3.8.2", etc]
+    ret = {}
+
+    for micro_version in sorted(python_versions, key = _ver_key):
+        minor_version, _, _ = micro_version.rpartition(".")
+
+        # This matches the raw flag value, e.g. --//python/config_settings:python_version=3.8
+        # It's private because matching the concept of e.g. "3.8" value is done
+        # using the `is_python_X.Y` config setting group, which is aware of the
+        # minor versions that could match instead.
+        ret.setdefault(minor_version, []).append(micro_version)
+
+        # Ensure that is_python_3.9.8 is matched if python_version is set
+        # to 3.9 if MINOR_MAPPING points to 3.9.8
+        default_micro_version = MINOR_MAPPING[minor_version]
+        ret[micro_version] = [minor_version] if default_micro_version == micro_version else []
+
+    return ret
+
+VERSION_FLAG_VALUES = _flag_values(TOOL_VERSIONS.keys())
+
+def is_python_config_setting(name, *, python_version, match_extra = None, **kwargs):
+    """Create a config setting for matching 'python_version' configuration flag.
+
+    This function is mainly intended for internal use within the `whl_library` and `pip_parse`
+    machinery.
+
+    Args:
+        name: name for the target that will be created to be used in select statements.
+        python_version: The python_version to be passed in the `flag_values` in the `config_setting`.
+        match_extra: The labels that should be used for matching the extra versions instead of creating
+            them on the fly. This will be passed to `config_setting_group.match_any`.
+        **kwargs: extra kwargs passed to the `config_setting`
+    """
+    visibility = kwargs.pop("visibility", [])
+
+    flag_values = {
+        _PYTHON_VERSION: python_version,
+    }
+    if python_version not in name:
+        fail("The name must have the python version in it")
+
+    match_extra = match_extra or {
+        "_{}".format(name).replace(python_version, x): {_PYTHON_VERSION: x}
+        for x in VERSION_FLAG_VALUES[python_version]
+    }
+    if not match_extra:
```

Comment:
I don't see how this block could ever trigger? VERSION_FLAG values will always be non-empty, so it'll always generate a non-empty match_extra dict.

Reply:
Changed the code semantics to use `match_any`.